In [70]:
import pandas as pd
import numpy as np
from pydataset import data

In [9]:
from env import host, user, password

def get_db_url(host,user,password):
    
    url = f'mysql+pymysql://{user}:{password}@{host}/employees'
    return url

In [12]:
url = get_db_url(host,user,password)
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)

emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

In [13]:
def get_db_url2(host,user,password):
    
    url = f'mysql+pym1ysql://{user}:{password}@{host}/employees'
    return url

In [46]:
url2 = get_db_url2(host,user,password)
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url2)  ##3 If we make a type we get a can't load plugin error

NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:mysql.pym1ysql

In [17]:
url = get_db_url(host,user,password)
pd.read_sql('SELECCT * FROM employees LIMIT 5 OFFSET 50', url) ##intentional error for error message review

ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'SELECCT * FROM employees LIMIT 5 OFFSET 50' at line 1")
[SQL: SELECCT * FROM employees LIMIT 5 OFFSET 50]
(Background on this error at: http://sqlalche.me/e/13/f405)

Question 4: Read the employees and titles tables into two separate DataFrames

In [22]:
employees_df = pd.read_sql('SELECT * FROM employees', url)
employees_df

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [19]:
type(employees_df)

pandas.core.frame.DataFrame

In [24]:
titles_df = pd.read_sql('SELECT * FROM titles', url)
titles_df

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

Question 5: How many rows and columns do you have in each dataframe?

In [25]:
print(employees_df.shape)
print(titles_df.shape)

(300024, 6)
(443308, 4)


Question 6: Display the summary statistics for each dataframe?

In [52]:
employees_df.describe(include='all')

emp_no  birth_date first_name last_name  gender   hire_date
count   300024.000000      300024     300024    300024  300024      300024
unique            NaN        4750       1275      1637       2        5434
top               NaN  1952-03-08     Shahab      Baba       M  1985-06-20
freq              NaN          95        295       226  179973         132
mean    253321.763392         NaN        NaN       NaN     NaN         NaN
std     161828.235540         NaN        NaN       NaN     NaN         NaN
min      10001.000000         NaN        NaN       NaN     NaN         NaN
25%      85006.750000         NaN        NaN       NaN     NaN         NaN
50%     249987.500000         NaN        NaN       NaN     NaN         NaN
75%     424993.250000         NaN        NaN       NaN     NaN         NaN
max     499999.000000         NaN        NaN       NaN     NaN         NaN

In [51]:
titles_df.describe(include='all')

emp_no     title   from_date     to_date
count   443308.000000    443308      443308      443308
unique            NaN         7        6393        5888
top               NaN  Engineer  1998-10-25  9999-01-01
freq              NaN    115003         132      240124
mean    253075.034430       NaN         NaN         NaN
std     161853.292613       NaN         NaN         NaN
min      10001.000000       NaN         NaN         NaN
25%      84855.750000       NaN         NaN         NaN
50%     249847.500000       NaN         NaN         NaN
75%     424891.250000       NaN         NaN         NaN
max     499999.000000       NaN         NaN         NaN

Question 7: How many unique titles are in the titles DataFrame

In [55]:
len(titles_df['title'].value_counts())

7

Question 8: What is the oldest_date in the to_date column?

In [42]:
titles_df['to_date'].sort_values().head(1) # Oldest to_date date below

16064    1985-03-01
Name: to_date, dtype: object

Question 9: What is the most recent date in the to_date column?

In [43]:
titles_df['to_date'].sort_values().tail(1) # Most recent to_date below

443307    9999-01-01
Name: to_date, dtype: object

In [44]:
min(titles_df['to_date']) ##some more wizardry

datetime.date(1985, 3, 1)

In [45]:
max(titles_df['to_date']) ##some more wizardry


datetime.date(9999, 1, 1)

In [56]:
pd.read_sql('SELECT COUNT(DISTINCT title) FROM titles', url)

COUNT(DISTINCT title)
0                      7

In [59]:
pd.read_sql("SELECT max(to_date) as most_recent_date FROM titles WHERE to_date NOT LIKE '9999%'", url) ## Using SQL to filter out the current 
## employees that have there to_date's set to 9999-01-01

most_recent_date
0       2002-08-01

## Exercises Part II Advanced Dataframes

In [60]:
users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [61]:
roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

Question 2: What is the result of using a right join on the DataFrames?

In [62]:
users.merge(roles, left_on='role_id', right_on='id', how='right', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only

Question 3: What is the result of using an outer join on the DataFrames?

In [63]:
users.merge(roles, left_on='role_id', right_on='id', how='outer', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

Question 4: What happens if you drop the foreign keys from the DataFrames and try to merge them?



In [64]:
users.merge(roles, how='outer', indicator=True)

id       name  role_id      _merge
0   1        bob      1.0   left_only
1   2        joe      2.0   left_only
2   3      sally      3.0   left_only
3   4       adam      3.0   left_only
4   5       jane      NaN   left_only
5   6       mike      NaN   left_only
6   1      admin      NaN  right_only
7   2     author      NaN  right_only
8   3   reviewer      NaN  right_only
9   4  commenter      NaN  right_only

# Load the mpg dataset from PyDataset

In [71]:
mpg = data('mpg')
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

For Question 7 we can see there are 234 rows and 11 columns after we load the data into our dataframe

In [73]:
## also can use
mpg.shape

(234, 11)

Question 8: Check out your column names and perform any cleanup you may want on them.

In [74]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




In [82]:
mpg = mpg.rename(columns={'cty': 'city', 'hwy': 'highway', 'displ': 'eng(L)', 'fuelType': 'fuel_type'})
mpg

manufacturer   model  eng(L)  year  cyl       trans drv  city  highway  \
1           audi      a4     1.8  1999    4    auto(l5)   f    18       29   
2           audi      a4     1.8  1999    4  manual(m5)   f    21       29   
3           audi      a4     2.0  2008    4  manual(m6)   f    20       31   
4           audi      a4     2.0  2008    4    auto(av)   f    21       30   
5           audi      a4     2.8  1999    6    auto(l5)   f    16       26   
..           ...     ...     ...   ...  ...         ...  ..   ...      ...   
230   volkswagen  passat     2.0  2008    4    auto(s6)   f    19       28   
231   volkswagen  passat     2.0  2008    4  manual(m6)   f    21       29   
232   volkswagen  passat     2.8  1999    6    auto(l5)   f    16       26   
233   volkswagen  passat     2.8  1999    6  manual(m5)   f    18       26   
234   volkswagen  passat     3.6  2008    6    auto(s6)   f    17       26   

    fuel_type    class  
1           p  compact  
2           p  compact  
3           p  compact  
4           p  compact  
5           p  compact  
..        ...      ...  
230         p  midsize  
231         p  midsize  
232         p  midsize  
233         p  midsize  
234         p  midsize  

[234 rows x 11 columns]

Question 9: Display the summary statistics for the dataset

In [84]:
mpg.describe(include = 'all')  #by using the include all in a describe we have it produce the summary statistics
# for all the columns

manufacturer        model      eng(L)         year         cyl  \
count           234          234  234.000000   234.000000  234.000000   
unique           15           38         NaN          NaN         NaN   
top           dodge  caravan 2wd         NaN          NaN         NaN   
freq             37           11         NaN          NaN         NaN   
mean            NaN          NaN    3.471795  2003.500000    5.888889   
std             NaN          NaN    1.291959     4.509646    1.611534   
min             NaN          NaN    1.600000  1999.000000    4.000000   
25%             NaN          NaN    2.400000  1999.000000    4.000000   
50%             NaN          NaN    3.300000  2003.500000    6.000000   
75%             NaN          NaN    4.600000  2008.000000    8.000000   
max             NaN          NaN    7.000000  2008.000000    8.000000   

           trans  drv        city     highway fuel_type class  
count        234  234  234.000000  234.000000       234   234  
unique        10    3         NaN         NaN         5     7  
top     auto(l4)    f         NaN         NaN         r   suv  
freq          83  106         NaN         NaN       168    62  
mean         NaN  NaN   16.858974   23.440171       NaN   NaN  
std          NaN  NaN    4.255946    5.954643       NaN   NaN  
min          NaN  NaN    9.000000   12.000000       NaN   NaN  
25%          NaN  NaN   14.000000   18.000000       NaN   NaN  
50%          NaN  NaN   17.000000   24.000000       NaN   NaN  
75%          NaN  NaN   19.000000   27.000000       NaN   NaN  
max          NaN  NaN   35.000000   44.000000       NaN   NaN

Question 10: How many different vehicle manufacturers are there?

In [85]:
mpg.manufacturer.value_counts().size 

15

In [94]:
mpg.manufacturer.agg('nunique')

15

In [95]:
mpg.manufacturer.nunique()

15

Question 11: How many different models are there?

In [97]:
mpg.model.value_counts().size

38

In [98]:
mpg.model.agg('nunique')

38

In [99]:
mpg.model.nunique()

38

Questions 12 & 13: Create mileage_difference columns and average_mileage columns for the vehicles

In [118]:
mpg['average_mileage2'] = mpg[['city','highway']].agg('mean', axis = 1)
mpg

manufacturer   model  eng(L)  year  cyl       trans drv  city  highway  \
1           audi      a4     1.8  1999    4    auto(l5)   f    18       29   
2           audi      a4     1.8  1999    4  manual(m5)   f    21       29   
3           audi      a4     2.0  2008    4  manual(m6)   f    20       31   
4           audi      a4     2.0  2008    4    auto(av)   f    21       30   
5           audi      a4     2.8  1999    6    auto(l5)   f    16       26   
..           ...     ...     ...   ...  ...         ...  ..   ...      ...   
230   volkswagen  passat     2.0  2008    4    auto(s6)   f    19       28   
231   volkswagen  passat     2.0  2008    4  manual(m6)   f    21       29   
232   volkswagen  passat     2.8  1999    6    auto(l5)   f    16       26   
233   volkswagen  passat     2.8  1999    6  manual(m5)   f    18       26   
234   volkswagen  passat     3.6  2008    6    auto(s6)   f    17       26   

    fuel_type    class  mileage_difference  average_mileage  is_automatic  \
1           p  compact                  11             23.5          True   
2           p  compact                   8             25.0         False   
3           p  compact                  11             25.5         False   
4           p  compact                   9             25.5          True   
5           p  compact                  10             21.0          True   
..        ...      ...                 ...              ...           ...   
230         p  midsize                   9             23.5          True   
231         p  midsize                   8             25.0         False   
232         p  midsize                  10             21.0          True   
233         p  midsize                   8             22.0         False   
234         p  midsize                   9             21.5          True   

     average_mileage2  
1                23.5  
2                25.0  
3                25.5  
4                25.5  
5                21.0  
..                ...  
230              23.5  
231              25.0  
232              21.0  
233              22.0  
234              21.5  

[234 rows x 15 columns]

In [100]:
mpg['mileage_difference'] = mpg.highway - mpg.city
mpg['average_mileage'] = (mpg.city + mpg.highway) / 2
mpg

manufacturer   model  eng(L)  year  cyl       trans drv  city  highway  \
1           audi      a4     1.8  1999    4    auto(l5)   f    18       29   
2           audi      a4     1.8  1999    4  manual(m5)   f    21       29   
3           audi      a4     2.0  2008    4  manual(m6)   f    20       31   
4           audi      a4     2.0  2008    4    auto(av)   f    21       30   
5           audi      a4     2.8  1999    6    auto(l5)   f    16       26   
..           ...     ...     ...   ...  ...         ...  ..   ...      ...   
230   volkswagen  passat     2.0  2008    4    auto(s6)   f    19       28   
231   volkswagen  passat     2.0  2008    4  manual(m6)   f    21       29   
232   volkswagen  passat     2.8  1999    6    auto(l5)   f    16       26   
233   volkswagen  passat     2.8  1999    6  manual(m5)   f    18       26   
234   volkswagen  passat     3.6  2008    6    auto(s6)   f    17       26   

    fuel_type    class  mileage_difference  average_mileage  
1           p  compact                  11             23.5  
2           p  compact                   8             25.0  
3           p  compact                  11             25.5  
4           p  compact                   9             25.5  
5           p  compact                  10             21.0  
..        ...      ...                 ...              ...  
230         p  midsize                   9             23.5  
231         p  midsize                   8             25.0  
232         p  midsize                  10             21.0  
233         p  midsize                   8             22.0  
234         p  midsize                   9             21.5  

[234 rows x 13 columns]

Question 14: Create a column named is_automatic determining if they are automatic or not

In [103]:
mpg['is_automatic'] = np.where(mpg.trans.str.count('auto'), 'automatic', 'manual')
mpg

manufacturer   model  eng(L)  year  cyl       trans drv  city  highway  \
1           audi      a4     1.8  1999    4    auto(l5)   f    18       29   
2           audi      a4     1.8  1999    4  manual(m5)   f    21       29   
3           audi      a4     2.0  2008    4  manual(m6)   f    20       31   
4           audi      a4     2.0  2008    4    auto(av)   f    21       30   
5           audi      a4     2.8  1999    6    auto(l5)   f    16       26   
..           ...     ...     ...   ...  ...         ...  ..   ...      ...   
230   volkswagen  passat     2.0  2008    4    auto(s6)   f    19       28   
231   volkswagen  passat     2.0  2008    4  manual(m6)   f    21       29   
232   volkswagen  passat     2.8  1999    6    auto(l5)   f    16       26   
233   volkswagen  passat     2.8  1999    6  manual(m5)   f    18       26   
234   volkswagen  passat     3.6  2008    6    auto(s6)   f    17       26   

    fuel_type    class  mileage_difference  average_mileage is_automatic  
1           p  compact                  11             23.5    automatic  
2           p  compact                   8             25.0       manual  
3           p  compact                  11             25.5       manual  
4           p  compact                   9             25.5    automatic  
5           p  compact                  10             21.0    automatic  
..        ...      ...                 ...              ...          ...  
230         p  midsize                   9             23.5    automatic  
231         p  midsize                   8             25.0       manual  
232         p  midsize                  10             21.0    automatic  
233         p  midsize                   8             22.0       manual  
234         p  midsize                   9             21.5    automatic  

[234 rows x 14 columns]

In [107]:
mpg['is_automatic'] = np.where(mpg.trans.str.count('auto'), True, False)
mpg

manufacturer   model  eng(L)  year  cyl       trans drv  city  highway  \
1           audi      a4     1.8  1999    4    auto(l5)   f    18       29   
2           audi      a4     1.8  1999    4  manual(m5)   f    21       29   
3           audi      a4     2.0  2008    4  manual(m6)   f    20       31   
4           audi      a4     2.0  2008    4    auto(av)   f    21       30   
5           audi      a4     2.8  1999    6    auto(l5)   f    16       26   
..           ...     ...     ...   ...  ...         ...  ..   ...      ...   
230   volkswagen  passat     2.0  2008    4    auto(s6)   f    19       28   
231   volkswagen  passat     2.0  2008    4  manual(m6)   f    21       29   
232   volkswagen  passat     2.8  1999    6    auto(l5)   f    16       26   
233   volkswagen  passat     2.8  1999    6  manual(m5)   f    18       26   
234   volkswagen  passat     3.6  2008    6    auto(s6)   f    17       26   

    fuel_type    class  mileage_difference  average_mileage  is_automatic  
1           p  compact                  11             23.5          True  
2           p  compact                   8             25.0         False  
3           p  compact                  11             25.5         False  
4           p  compact                   9             25.5          True  
5           p  compact                  10             21.0          True  
..        ...      ...                 ...              ...           ...  
230         p  midsize                   9             23.5          True  
231         p  midsize                   8             25.0         False  
232         p  midsize                  10             21.0          True  
233         p  midsize                   8             22.0         False  
234         p  midsize                   9             21.5          True  

[234 rows x 14 columns]

Question 15: Using the mpg dataset, find out which manufacturer has the best miles per gallon on average?

In [111]:
mpg.groupby('manufacturer').average_mileage.agg('mean').sort_values(ascending=False)

manufacturer
honda         28.500000
volkswagen    25.074074
hyundai       22.750000
subaru        22.428571
audi          22.027778
toyota        21.720588
pontiac       21.700000
nissan        21.346154
chevrolet     18.447368
ford          16.680000
mercury       15.625000
jeep          15.562500
dodge         15.540541
lincoln       14.166667
land rover    14.000000
Name: average_mileage, dtype: float64

Question 16: Do automatic or manual cars have better miles per gallon?

In [114]:
mpg.groupby('is_automatic').average_mileage.agg('mean').sort_values() #Looks like manual cars are the winners here

is_automatic
True     19.130573
False    22.227273
Name: average_mileage, dtype: float64